src/engines/drum_live.c

```diff
@@ -1,149 +1,920 @@
+        dest[7] = src[7] & mask;
+        dest += 8;
+        src += 8;
+    }
+}
+
+
+// Multiply Palette
+void drum_live_multiply_palette(const u16 *src, u16 *dest, u32 totalPal, u32 brgTargets, u32 mul) {
+    u32 blueMul, redMul, greenMul;
+    u32 blueVal, redVal, greenVal;
+    u32 totalColors;
+    u32 i;
+
+    // 'mul' is a 5-bit fixed-point integer.
+    blueMul = (brgTargets & 1) ? 0x20 : mul;
+    redMul = ((brgTargets >> 1) & 1) ? 0x20 : mul;
+    greenMul = ((brgTargets >> 2) & 1) ? 0x20 : mul;
+    totalColors = totalPal * 16;
+
+    for (i = 0; i < totalColors; i++) {
+        u32 color = *src++;
+
+        blueVal = (color >> 10) & 0x1F;
+        redVal = (color >> 0) & 0x1F;
+        greenVal = (color >> 5) & 0x1F;
+        blueVal = (blueVal * blueMul) >> 5;
+        redVal = (redVal * redMul) >> 5;
+        greenVal = (greenVal * greenMul) >> 5;
+
+        *dest++ = ((blueVal << 10) | redVal | (greenVal << 5));
+    }
+}
+
+
+// Update Rainbow Lighting
+void drum_live_update_rainbow_lights(void) {
+    u32 brgTargets;
+
+    gDrumLive->nextLightBRG = (gDrumLive->nextLightBRG + 1) & 3;
+    brgTargets = drum_live_brg_targets[gDrumLive->nextLightBRG];
+    drum_live_multiply_palette(drum_live_obj_palettes[gDrumLive->version][0], gDrumLive->objPalBuf4, 4, brgTargets, 0x12);
+    drum_live_multiply_palette(drum_live_bg_palettes[gDrumLive->version][0], gDrumLive->bgPalBuf4, 4, brgTargets, 0x1A);
+}
+
+
+// Lighting Beat Animation
+void drum_live_flash_beat_lights(void) {
+    dma3_set(gDrumLive->objPalBuf4, gDrumLive->objPalBuf3, 0x80, 0x20, 0x200);
+    dma3_set(gDrumLive->bgPalBuf4, gDrumLive->bgPalBuf3, 0x80, 0x20, 0x200);
+
+    switch (gDrumLive->currentLightStyle) {
+        case 0:
+            dma3_set(drum_live_obj_palettes[gDrumLive->version], gDrumLive->objPalBuf4, 0x80, 0x20, 0x200);
+            dma3_set(drum_live_bg_palettes[gDrumLive->version], gDrumLive->bgPalBuf4, 0x80, 0x20, 0x200);
+            break;
+
+        case 1:
+            drum_live_update_rainbow_lights();
+            break;
+    }
+
+    func_08001ddc(0x10, 4, gDrumLive->objPalBuf3, gDrumLive->objPalBuf4, gDrumLive->objPalBuf3);
+    func_08001ddc(0x10, 4, gDrumLive->bgPalBuf3, gDrumLive->bgPalBuf4, gDrumLive->bgPalBuf3);
+    func_08001fe0(get_current_mem_id(), ticks_to_frames(4), 4, gDrumLive->objPalBuf3, gDrumLive->objPalBuf4, gDrumLive->objPalBuf1);
+    func_08001fe0(get_current_mem_id(), ticks_to_frames(4), 4, gDrumLive->bgPalBuf3, gDrumLive->bgPalBuf4, gDrumLive->bgPalBuf1);
+}
+
+
+// Engine Event 0x06 (Set Next Light Style)
+void drum_live_set_next_beat_light_style(u32 style) {
+    gDrumLive->nextLightStyle = style;
+    gDrumLive->timeUntilNextLightStyle = ticks_to_frames(20);
+}
+
+
+// Engine Event 0x07 (Play Applause)
+void drum_live_script_play_applause(void) {
+    drum_live_play_applause();
+}
+
+
+// Clear Cool/Lame Inputs
+void drum_live_clear_input_def(void) {
+    u32 i;
+
+    for (i = 0; i < 6; i++) {
```

Comment:
Add a define for the amount of `coolInputs` and `lameInputs` like `#define DRUM_LIVE_INPUT_AMOUNT 6` and 
```suggestion
    for (i = 0; i < DRUM_LIVE_INPUT_AMOUNT; i++) {
```